velox/functions/lib/Re2Functions.cpp

```diff
@@ -424,30 +490,101 @@ bool matchSubstringPattern(
       std::string::npos);
 }
 
+// For pattern '___hello___', it is split into three sub-patterns:
+//   - kSingleCharWildcard: ___
+//   - kLiteralString: hello
+//   - kSingleCharWildcard: ___
+//
+// A naive implementation is search the three pattern every time, a better
+// implementation is: skip the length of leading kSingleCharWildcard chars(3
+// here), and search the first kLiteralString string in input to find the
+// position to match.
+bool matchRelaxedSubstringPattern(
+    StringView input,
+    const PatternMetadata& patternMetadata) {
+  if (input.size() < patternMetadata.length()) {
+    return false;
+  }
+
+  // Calculate the necessary search range: [start, end].
+  auto start = patternMetadata.leadingSingleWildcardCharsCount();
+  auto startPatternIndex = start > 0 ? 1 : 0;
+  auto firstFixedString =
+      patternMetadata.subPatterns()[startPatternIndex].pattern.c_str();
+  auto indexInString = start;
+  auto end = input.size() - patternMetadata.restFixedLength() + start;
+
+  while (indexInString <= end) {
+    // Search the firstFixedString to find out where to start match the whole
+    // pattern.
+    auto it = strstr(input.begin() + indexInString, firstFixedString);
```

Comment:
Is it safe to apply strstr to non-zero-terminated strings. I feel this might not be safe. I'm also concerned about the performance of some edge cases where we need to keep running this loop many time shifting one char at a time. Would it be OK to not support this pattern, at least not in this PR, and focus on simple prefix and suffix patterns?